### PR TITLE
executor: refineArgs() bug fix when compare int with very small decimal

### LINKED ISSUE #23623: wrong result when including subquery
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
use test;
source [test.txt](https://github.com/pingcap/tidb/files/6219774/test.txt);
select sum(col1) from UK_SIGNED_19370 t1 where t1.col1 > -2147483648;
select count(*) from UK_SIGNED_19370 t1 where t1.col1 > -14955637448;
select count(*) from UK_SIGNED_19370 t1 where t1.col1 > (select sum(col1) from UK_SIGNED_19370 t1 where t1.col1 > -2147483648);
```

### 2. What did you expect to see? (Required)
```
mysql> select sum(col1) from UK_SIGNED_19370 t1 where t1.col1 > -2147483648;
+--------------+
| sum(col1)    |
+--------------+
| -14955637448 |
+--------------+
1 row in set (0.05 sec)

mysql> select count(*) from UK_SIGNED_19370 t1 where t1.col1 > -14955637448;
+----------+
| count(*) |
+----------+
|      102 |
+----------+
1 row in set (0.06 sec)

mysql> select count(*) from UK_SIGNED_19370 t1 where t1.col1 > (select sum(col1) from UK_SIGNED_19370 t1 where t1.col1 > -2147483648);
+----------+
| count(*) |
+----------+
|      102 |
+----------+
1 row in set (0.05 sec)
```

### 3. What did you see instead (Required)
```
mysql> select sum(col1) from UK_SIGNED_19370 t1 where t1.col1 > -2147483648;
+--------------+
| sum(col1)    |
+--------------+
| -14955637448 |
+--------------+
1 row in set (0.00 sec)

mysql> select count(*) from UK_SIGNED_19370 t1 where t1.col1 > -14955637448;
+----------+
| count(*) |
+----------+
|      102 |
+----------+
1 row in set (0.00 sec)

mysql> select count(*) from UK_SIGNED_19370 t1 where t1.col1 > (select sum(col1) from UK_SIGNED_19370 t1 where t1.col1 > -2147483648);
+----------+
| count(*) |
+----------+
|      103 |
+----------+
1 row in set (0.00 sec)
```
### 4. What is your TiDB version? (Required)

master v4.0.0-beta.2-2479-g1e3921aff
release-5.0 v5.0.0-nightly-51-g638272efb
release-4.0 v4.0.11-40-g11a9254be

### LINKED ISSUE #23684: executor: wrong results from subquery
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
mysql -h 172.16.5.85 -P 53324 -u root -D agg_sub_datat4000


```
SELECT count(*),  table1 . `col_decimal_40_not_null_key` <= ( SELECT   SUM(  SUBQUERY1_t1 . `col_decimal_not_null_key` ) AS SUBQUERY1_field1 FROM ( `GG` AS SUBQUERY1_t1 STRAIGHT_JOIN `M` AS SUBQUERY1_t2 ON (SUBQUERY1_t2 . `col_varchar_binary_key` = SUBQUERY1_t1 . `col_varchar_binary`  ) )  ) AS field1 ,  table1 . `col_varchar_binary_not_null_key` != ( SELECT 'i' FROM DUAL ) AS field2 FROM ( `Q` AS table1 INNER JOIN `K` AS table2 ON (table2 . `col_varchar_binary_key` = table1 . `col_varchar_binary_not_null_key`  ) ) WHERE (  table1  . `col_int_key` <= ( SELECT DISTINCT  SUM(  SUBQUERY3_t1 . `col_int` ) AS SUBQUERY3_field1 FROM `K` AS SUBQUERY3_t1 WHERE SUBQUERY3_t1 . `col_decimal_key` > 4 ) ) OR table1 . `col_varchar_binary_key`  IN ('u', 'm') AND ( table1 . `col_varchar_binary` < 'z' AND table1 . `col_varchar_binary_not_null` <> 'x' )   group by 2,3;
```
mysql -h 172.16.5.85 -P 53324 -u root -D agg_sub_datat10000
```
SELECT  ( 'p', 'p' )  IN ( SELECT   SUBQUERY1_t1 . `col_varchar_binary` AS SUBQUERY1_field1 , SUBQUERY1_t1 . `col_varchar_binary_not_null_key` AS SUBQUERY1_field2 FROM `V` AS SUBQUERY1_t1    ) AS field1 FROM ( `L` AS table1 STRAIGHT_JOIN ( ( `JJ` AS table2 INNER JOIN `B` AS table3 ON (table3 . `col_decimal_40` = table2 . `col_int_key`  ) ) ) ON (table3 . `pk` = table2 . `col_int_not_null`  ) ) WHERE (  table3 . `col_int_key` < ( SELECT   SUM(  SUBQUERY2_t1 . `col_int_key` ) AS SUBQUERY2_field1 FROM ( `B` AS SUBQUERY2_t1 RIGHT  JOIN `FF` AS SUBQUERY2_t2 ON (SUBQUERY2_t2 . `col_int_key` = SUBQUERY2_t1 . `col_decimal_40`  ) )  ) ) AND table1 . `col_varchar_binary_key` = table3 . `col_varchar_binary_key` AND ( table1 . `col_varchar_binary_not_null` IS NOT NULL OR table1 . `col_decimal_40_key` = table1 . `col_decimal_40_not_null_key` )  GROUP BY field1  ;
```
<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

on 172.16.5.85 mysql
mysql -h 127.0.0.1 -P 3306 -u root -D agg_sub_datat10000
```
+----------+--------+--------+
| count(*) | field1 | field2 |
+----------+--------+--------+
|        1 |      1 |      0 |
|       18 |      1 |      1 |
+----------+--------+--------+
```

```
empty
```
### 3. What did you see instead (Required)
```
mysql> set @@tidb_isolation_read_engines='tikv,tidb'; set @@tidb_allow_mpp=0;
Query OK, 0 rows affected (0.05 sec)

Query OK, 0 rows affected (0.03 sec)

mysql> SELECT count(*),  table1 . `col_decimal_40_not_null_key` <= ( SELECT   SUM(  SUBQUERY1_t1 . `col_decimal_not_null_key` ) AS SUBQUERY1_field1 FROM ( `GG` AS SUBQUERY1_t1 STRAIGHT_JOIN `M` AS SUBQUERY1_t2 ON (SUBQUERY1_t2 . `col_varchar_binary_key` = SUBQUERY1_t1 . `col_varchar_binary`  ) )  ) AS field1 ,  table1 . `col_varchar_binary_not_null_key` != ( SELECT 'i' FROM DUAL ) AS field2 FROM ( `Q` AS table1 INNER JOIN `K` AS table2 ON (table2 . `col_varchar_binary_key` = table1 . `col_varchar_binary_not_null_key`  ) ) WHERE (  table1  . `col_int_key` <= ( SELECT DISTINCT  SUM(  SUBQUERY3_t1 . `col_int` ) AS SUBQUERY3_field1 FROM `K` AS SUBQUERY3_t1 WHERE SUBQUERY3_t1 . `col_decimal_key` > 4 ) ) OR table1 . `col_varchar_binary_key`  IN ('u', 'm') AND ( table1 . `col_varchar_binary` < 'z' AND table1 . `col_varchar_binary_not_null` <> 'x' )   group by 2,3;
+----------+--------+--------+
| count(*) | field1 | field2 |
+----------+--------+--------+
|       22 |      1 |      1 |
|        2 |      1 |      0 |
+----------+--------+--------+
2 rows in set (0.06 sec)

mysql> desc SELECT count(*),  table1 . `col_decimal_40_not_null_key` <= ( SELECT   SUM(  SUBQUERY1_t1 . `col_decimal_not_null_key` ) AS SUBQUERY1_field1 FROM ( `GG` AS SUBQUERY1_t1 STRAIGHT_JOIN `M` AS SUBQUERY1_t2 ON (SUBQUERY1_t2 . `col_varchar_binary_key` = SUBQUERY1_t1 . `col_varchar_binary`  ) )  ) AS field1 ,  table1 . `col_varchar_binary_not_null_key` != ( SELECT 'i' FROM DUAL ) AS field2 FROM ( `Q` AS table1 INNER JOIN `K` AS table2 ON (table2 . `col_varchar_binary_key` = table1 . `col_varchar_binary_not_null_key`  ) ) WHERE (  table1  . `col_int_key` <= ( SELECT DISTINCT  SUM(  SUBQUERY3_t1 . `col_int` ) AS SUBQUERY3_field1 FROM `K` AS SUBQUERY3_t1 WHERE SUBQUERY3_t1 . `col_decimal_key` > 4 ) ) OR table1 . `col_varchar_binary_key`  IN ('u', 'm') AND ( table1 . `col_varchar_binary` < 'z' AND table1 . `col_varchar_binary_not_null` <> 'x' )   group by 2,3;
+-------------------------------------+---------+-----------+--------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| id                                  | estRows | task      | access object                                                      | operator info                                                                                                                                                                                                                                     |
+-------------------------------------+---------+-----------+--------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Projection_148                      | 15.36   | root      |                                                                    | Column#271, le(agg_sub_datat1000.q.col_decimal_40_not_null_key, 63)->Column#347, ne(agg_sub_datat1000.q.col_varchar_binary_not_null_key, i)->Column#350                                                                                           |
| └─HashAgg_149                       | 15.36   | root      |                                                                    | group by:Column#361, Column#362, funcs:count(1)->Column#271, funcs:firstrow(Column#359)->agg_sub_datat1000.q.col_varchar_binary_not_null_key, funcs:firstrow(Column#360)->agg_sub_datat1000.q.col_decimal_40_not_null_key                         |
|   └─Projection_194                  | 24.00   | root      |                                                                    | agg_sub_datat1000.q.col_varchar_binary_not_null_key, agg_sub_datat1000.q.col_decimal_40_not_null_key, le(agg_sub_datat1000.q.col_decimal_40_not_null_key, 63)->Column#361, ne(agg_sub_datat1000.q.col_varchar_binary_not_null_key, i)->Column#362 |
|     └─HashJoin_177                  | 24.00   | root      |                                                                    | inner join, equal:[eq(agg_sub_datat1000.q.col_varchar_binary_not_null_key, agg_sub_datat1000.k.col_varchar_binary_key)]                                                                                                                           |
|       ├─TableReader_187(Build)      | 19.20   | root      |                                                                    | data:Selection_186                                                                                                                                                                                                                                |
|       │ └─Selection_186             | 19.20   | cop[tikv] |                                                                    | or(1, and(in(agg_sub_datat1000.q.col_varchar_binary_key, "u", "m"), and(lt(agg_sub_datat1000.q.col_varchar_binary, "z"), ne(agg_sub_datat1000.q.col_varchar_binary_not_null, "x"))))                                                              |
|       │   └─TableFullScan_185       | 24.00   | cop[tikv] | table:table1                                                       | keep order:false, stats:pseudo                                                                                                                                                                                                                    |
|       └─IndexReader_193(Probe)      | 21.98   | root      |                                                                    | index:IndexFullScan_192                                                                                                                                                                                                                           |
|         └─IndexFullScan_192         | 21.98   | cop[tikv] | table:table2, index:col_varchar_binary_key(col_varchar_binary_key) | keep order:false, stats:pseudo                                                                                                                                                                                                                    |
+-------------------------------------+---------+-----------+--------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
9 rows in set (0.05 sec)
```
```
+--------+
| field1 |
+--------+
|      0 |
+--------+
```
### 4. What is your TiDB version? (Required)
master
<!-- Paste the output of SELECT tidb_version() -->

### PR DESCRIPTION
**Pull request by guo-shaoge**
Signed-off-by: guo-shaoge <shaoge1994@163.com>

### What problem does this PR solve?

Issue Number: 
close #23623 
close #23684 

Problem Summary: In the following case, output of select stmt should be 2, but current result is 3. 

Why: c1 is nullable, so we cannot assuse all value in c1 will be greater than the result of subquery, which is less than int32_min. Because result of `null > xxx` is NULL

    drop table if exists t1;
    create table t1(c1 int);
    insert into t1 values(-2147483648), (-2147483648), (null);
    select count(*) from t1 where c1 > (select sum(c1) from t1);

### What is changed and how it works?

What's Changed: when `refineArgs()`, should check if column is nullable. If is nullable, we should not try to optimize compare expression.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects

- Performance regression
    - Consumes more CPU: may cause performance regression compared when compare int_column with const if int_column is nullable.

### Release note 

- fix bug when compare int_column with constant value

### CODE REVIEW

File: expression/builtin_compare.go [hunk 1]
```diff
@@ -1351,7 +1351,10 @@ func (c *compareFunctionClass) refineArgs(ctx sessionctx.Context, args []Express
...
 	isPositiveInfinite, isNegativeInfinite := false, false
 	// int non-constant [cmp] non-int constant
-	if arg0IsInt && !arg0IsCon && !arg1IsInt && arg1IsCon {
+    // Why check not null flag
+    // eg: int_col > const_val(which is less than min_int32)
+    // If int_col got null, compare result cannot be true
+	if arg0IsInt && !arg0IsCon && !arg1IsInt && arg1IsCon && mysql.HasNotNullFlag(arg0Type.Flag) {
 		arg1, isExceptional = RefineComparedConstant(ctx, *arg0Type, arg1, c.op)
 		finalArg1 = arg1
 		if isExceptional && arg1.GetType().EvalType() == types.ETInt {
```
Comment (XuHuaiyu) [inline expression/builtin_compare.go:1360]: If `arg1 == false`, can we ignore the not-null flag ?

**Comment by XuHuaiyu** [comment]
CI needs to be fixed.

File: expression/builtin_compare.go (hunk 1 above)
Comment (guo-shaoge) [reply expression/builtin_compare.go]: select xxx where (col_int > (select sum(xx) xx) ) is null;

So need to distinguish null or false of `> expr`.

**Pull request merged**